src/typeChecker/typeErrorChecker.ts

```diff
@@ -0,0 +1,856 @@
+      const expectedParamTypes = expectedType.parameterTypes
+      if (actualParamTypes.length !== expectedParamTypes.length) {
+        return true
+      }
+      for (let i = 0; i < actualType.parameterTypes.length; i++) {
+        // Note that actual and expected types are swapped here
+        // as expected type should be subset of actual type for function arguments
+        if (hasTypeMismatchErrors(expectedParamTypes[i], actualParamTypes[i])) {
+          return true
+        }
+      }
+      // Check return type
+      return hasTypeMismatchErrors(actualType.returnType, expectedType.returnType)
+    case 'union':
+      const expectedSet = new Set(expectedType.types.map(formatTypeString))
+      // If not union type, expected set should contain actual type
+      if (actualType.kind !== 'union') {
+        return !expectedSet.has(formatTypeString(actualType))
+      }
+      // If both are union types, actual set should be a subset of expected set
+      const actualSet = new Set(actualType.types.map(formatTypeString))
+      for (const elem of actualSet) {
+        if (!expectedSet.has(elem)) {
+          return true
+        }
+      }
+      return false
+    default:
+      return true
+  }
+}
+
+/**
+ * Checks the types of the arguments of the given call expression.
+ * If number of arguments is different, add InvalidNumberOfArguments error to context and terminates early.
+ * Else, checks each argument against its expected type.
+ */
+function checkArgTypes(
+  node: es.CallExpression,
+  expectedTypes: Primitive[],
+  context: Context,
+  env: TypeEnvironment
+) {
+  const args = node.arguments
+  if (args.length !== expectedTypes.length) {
+    context.errors.push(new InvalidNumberOfArguments(node, expectedTypes.length, args.length))
+    return
+  }
+  for (let i = 0; i < expectedTypes.length; i++) {
+    const expectedType = expectedTypes[i]
+    const node = args[i]
+    const actualType = typeCheckAndReturnType(node, context, env)
+    checkForTypeMismatch(node, actualType, expectedType, context)
+  }
+}
+
+/**
+ * Converts type annotation node to its corresponding type representation in Source.
+ * If no type annotation exists, returns the "any" primitive type.
+ */
+function getAnnotatedType(
+  annotationNode: AnnotationTypeNode | TypeAliasDeclarationNode | undefined,
+  context: Context,
+  env: TypeEnvironment
+): Type {
+  if (!annotationNode) {
+    return tAny
+  }
+  const annotatedTypeNode = annotationNode.typeAnnotation
+  switch (annotatedTypeNode.type) {
+    case TSNodeType.TSFunctionType:
+      const fnTypeNode = annotatedTypeNode as FunctionTypeNode
+      const fnTypes = getParamTypes(fnTypeNode.parameters, context, env)
+      // Return type will always be last item in types array
+      fnTypes.push(getAnnotatedType(fnTypeNode.typeAnnotation, context, env))
+      return tFunc(...fnTypes)
+    case TSNodeType.TSUnionType:
+      const unionTypeNode = annotatedTypeNode as UnionTypeNode
```

Comment:
Same here, the typecasting wouldn't be necessary if discriminated unions were used.